### DISCUSSION TITLE
Modifying 'Tabbed Window'

### DISCUSSION THREAD

**Discussion by Daemoen** [scandum/tintin#116] [Script help]
Lo All!

Started getting back into tinkering with Tintin, it's come a long long ways since I last tried to do anything fancy with it, I am impressed!

I looked through the scripts and came across tabbed window, this is almost perfect for what I want to do... except I want more tabs, I want to pad the tabs a bit, and I want to output to more than 1 tab at a time as well....

So first things first, I added the new tabs (all, activities, says) just fine, as well as the unread counters...

Then I changed the box to suit my needs, but now I am attempting to change the comm_window_draw_tabs and a bit stuck:

```
#ALIAS {comm_window_draw_tabs}
{
    #draw green rounded box @comm_window_border{};
    #local short {};
    #local index {};
    #loop {1} {&comm_window[tab][]} {index}
    {
        #local tab *comm_window[tab][+$index];
        #if {{$comm_window[active]} === {$tab}}
        {
            #format short <138>%+10s $tab
        };
        #elseif {$comm_window[unread][$tab] == 1}
        {
            #format short <168>%+10s $tab
        };
        #else
        {
            #format short <268>%+10s $tab
        };
        #replace short { } {<028>-};
        #line ignore #showme {\e]68;2;TAB;comm_window_tab_click $tab\a\e[4;24m$short\e[24m} {$comm_window[toprow]} {$comm_window[topcol] - 9 + $index * 11}
    }
}
```

I added the #replace short {} {} myself, but I also understand why it's not working quite how I want...  With the way that it works above, when it replaces the padded -s with the green colored ones, it of course overwrites the formatting specified right above it in the if....  Is there a better way to pad the format's in this case that I can set the -s to green to match the border, while retaining the remainder of the color features from the active tab/unread tab color settings?  I didn't see an option for that in format, or any other simple way of doing it.

The next part is a bit more tricky, based on how I believe the script works:
```
#ACTION {~%1 tells you %2}
{
    comm_window_show tells %1 tells you %2
}
```

I would like to have it actually do comm_window_show all %1 tells you %2 (and use this in all of my other primary capture triggers), but also have it store the tells for display in tells, chats in display for chats, etc...

Because of the way it works in that script, I don't think it will be possible, without some pretty massive changes to the script.  I had considered that maybe I could initialize a variable like the way worldlst works, and use that to store a given number of the most recent messages for any given channel, displaying them in a tab based on the content of that variable... but not sure if that would be the best way of going about it, or if there is something simpler/better that I am overlooking.

Any input/suggestions welcome.

Thanks!

**Comment by Daemoen** [comment]
I went back and looked through the script a bit more, it already does exactly what I need in order to store the messages in the various lists (since it uses #list to expand $comm_window[tab][*]), so I guess it won't be quite as much work as I feared

**Comment by scandum** [comment]
Wouldn't it be enough to change
```
#format short <138>%+10s $tab
```
to
```
#format short <138>%+10s<028>-$tab
```
?

You could use:
```
#ACTION {~%1 tells you %2}
{
    comm_window_show all %1 tells you %2;
    comm_window_show tells %1 tells you %2
}
```
You could also update the comm_window_show alias to automatically add everything to the 'all' tab.

**Comment by Daemoen** [comment]
Been rather busy, so didn't get back to this until now...  noticed a few things..

* If I go afk for a while and leave my session running in byobu like usual, eventually I come back and the fancy box is gone, until I force it to redraw... not sure what information you need to even begin to troubleshoot or understand how/why that's happening, but it is very much repeatable.
* I have attempted a couple of basic triggers, and they 'kind of' work... but they seem to strip color, no matter how I go about capturing it...

Ex:
On the mud, the following text is in bold cyan:
[chat] You: test2
[chat] Anatak: You mean 3

I have tried numerous formats for the actions:

```
#action {~^\[chat\] %1} {comm_window_show 1-chats %0; comm_window_show 0-chats %0}
#action {~\[chat\] {%*}$} {comm_window_show 1-chats %0; comm_window_show 0-chats %0}
#action {^[chat] %1} {comm_window_show 1-chats %0; comm_window_show 0-chats %0}
#action {~^[chat] {%*}$} {comm_window_show 1-chats %0; comm_window_show 0-chats %0}
```
(and other variations of the above, I'm just not going to write them all)

My requirements are pretty simple, but the way that triggering in Tintin still doesn't make a whole lot of sense to me.  it seems the %1 and similar capture groups are 'greedy capture' groups somehow, though I haven't dug into the code to find out exactly how it works.  

Is there a way of using *pure regex* for the triggers, while also retaining the color of the matched text?  (is the color stored when captured into %0?)

**Comment by scandum** [comment]
You can use `#config convert on` to see the color codes used by the mud, those need to be properly matched when using ~.

You can embed a pure regex using `#act {{<regex>}}`, so just an extra set of braces.
